### ISSUE TITLE
TwitchChatClient onTimeout does not provide reason

### ISSUE THREAD

**Issue by lafiosca** [d-fischer/twitch#37]
## Bug Report

Assuming that `onTimeout` worked in the past, it appears to me that something has changed about the chat message format that communicates chat timeouts. The `onTimeout` method does not provide a value for `reason` (the third argument), and the raw output from the underlying IRC client does not appear to contain the timeout reason.

### Code

Nothing fancy, just set up a TwitchChatClient on a channel and listen for timeout events:

```ts
twitchChat.onTimeout((channel, user, reason, duration) => {
    console.log(JSON.stringify({
        channel,
        user,
        reason,
        duration,
    }));
})
```

Or hook directly into the ChatClear message like the library does:

```ts
import ClearChat from 'twitch-chat-client/lib/Capabilities/TwitchCommands/MessageTypes/ClearChat';

twitchChat.onMessage(ClearChat, (args) => {
    console.log(`CLEAR CHAT: ${util.inspect(args, { showHidden: true, depth: null })}`);
});
```

### Expected behavior

When I type `/timeout foobar 10 this is a test` into Twitch chat, I expect the `onTimeout` event to be called with a string value of `'this is a test'` for `reason`.

### Actual Behavior

Instead, `reason` seems to be undefined.

Looking at the [code](https://github.com/d-fischer/twitch/blob/5f9045e7527324600a6dfe62ad820c59403df16f/packages/twitch-chat-client/src/ChatClient.ts#L483), it seems the reason was previously stored in a tag called `ban-reason`, but I do not see this tag when testing:

```
ClearChat {
  _params:
   [ { value: '#lafiosca', trailing: false },
     { value: 'foobar', trailing: true },
     [length]: 2 ],
  _serverProperties:
   { channelTypes: '#&',
     supportedUserModes: 'iwso',
     supportedChannelModes:
      { prefix: 'ov',
        list: 'b',
        alwaysWithParam: 'ovk',
        paramWhenSet: 'l',
        noParam: 'imnpst' },
     prefixes:
      [ { modeChar: 'v', prefix: '+' },
        { modeChar: 'o', prefix: '@' },
        [length]: 2 ] },
  _command: 'CLEARCHAT',
  _tags:
   Map {
     'ban-duration' => '10',
     'room-id' => '81918359',
     'target-user-id' => '16231',
     'tmi-sent-ts' => '1555130783309',
     [size]: 4 },
  _prefix: { nick: 'tmi.twitch.tv' },
  _raw: '@ban-duration=10;room-id=81918359;target-user-id=16231;tmi-sent-ts=1555130783309 :tmi.twitch.tv CLEARCHAT #lafiosca :foobar',
  _parsedParams:
   { channel: { value: '#lafiosca', trailing: false },
     user: { value: 'foobar', trailing: true } } }
```

I wonder if it is even still possible for us to get this value. The Twitch dashboard chat client seems to show the message (at least for the person who runs the timeout command), which makes me think it might be communicated some other way now.

**Comment by freaktechnik** [comment]
Twitch removed the reason tag recently. You can still get it in PubSub (the moderation topic that I think the pubsub lib recently got support for). The reason the tag is still in the type is most likely for backward comapt.

**Comment by lafiosca** [comment]
Per discussion on Discord, @d-fischer is indeed leaving it in there for backwards compatibility. Although technically I think backwards compatibility is already being broken in a sense; `reason` is supposed to be a `string`, not `undefined`. (Which is not to imply that returning `''` would be any better!)

Tangentially, perhaps upon a future major version (breaking change) these handler signatures could be changed to use bags of parameters rather than ordered argument lists which are prone to this kind of breakage. For example,

```
onTimeout((channel: string, user: string, reason: string, duration: number) => void)
```

could be replaced with something like

```
interface TimeoutEvent {
  channel: string;
  user: string;
  duration?: number;
}

onTimeout(timeoutEvent: TimeoutEvent => void)
```

This approach assumes that a timeout event always includes a channel and user (because it would have no meaning otherwise) and that the duration is additional qualitative data. Then if duration were removed from the tags in a future update this typing would still work, with the library providing an `undefined` duration. If other tags were added in future updates, these could also easily be added to the bag interface without having to consider the order of an ever-growing argument list. This last point could be especially relevant to the numerous methods that include the raw message object as their final argument.

This approach is not without its flaws, though it also allows for a more "buffet-style" approach to picking from the provided values when they are not all relevant to the consumer. To wit, if one of the bag values was removed due to an API change, and I had not been using it at all, I would never even notice.

**Comment by lafiosca** [comment]
I am going to close the issue now, though it might be revisiting if a breaking change release is on the horizon, at least to remove the `reason` argument if nothing else.

**Comment by d-fischer** [comment]
Fixed in 944ee4f and released as 4.0.0-pre.0.
